compiler/rustc_lint/src/lifetime_style.rs

```diff
@@ -0,0 +1,454 @@
+use rustc_data_structures::fx::FxIndexMap;
+use rustc_hir::intravisit::{self, Visitor};
+use rustc_hir::{self as hir, LifetimeSource, LifetimeSyntax};
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::Span;
+use tracing::instrument;
+
+use crate::{LateContext, LateLintPass, LintContext, lints};
+
+declare_lint! {
+    /// The `mismatched_lifetime_syntaxes` lint detects when an
+    /// elided lifetime uses different syntax between function
+    /// arguments and return values.
+    ///
+    /// The three kinds of syntax are:
+    /// 1. Explicitly bound lifetimes, such as `'static` or `'a`.
+    /// 2. The explicitly anonymous lifetime `'_`.
+    /// 3. Implicit lifetimes, such as `&u8` or `ThisHasALifetimeGeneric`.
+    ///
+    /// As an exception, this lint allows references with implicit or
+    /// explicitly anonymous lifetimes to be paired with paths using
+    /// explicitly anonymous lifetimes.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(mismatched_lifetime_syntaxes)]
+    ///
+    /// pub fn mixing_explicit_bound_with_implicit(v: &'static u8) -> &u8 {
+    ///     v
+    /// }
+    ///
+    /// struct Person<'a> {
+    ///     name: &'a str,
+    /// }
+    ///
+    /// pub fn mixing_implicit_with_explicit_anonymous(v: Person) -> Person<'_> {
+    ///     v
+    /// }
+    ///
+    /// struct Foo;
+    ///
+    /// impl Foo {
+    ///     // Lifetime elision results in the output lifetime becoming
+    ///     // `'static`, which is not what was intended.
+    ///     pub fn get_mut(&'static self, x: &mut u8) -> &mut u8 {
+    ///         unsafe { &mut *(x as *mut _) }
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Lifetime elision is useful because it frees you from having to
+    /// give each lifetime its own name and show the relation of input
+    /// and output lifetimes for common cases. However, a lifetime
+    /// that uses inconsistent syntax between related arguments and
+    /// return values is more confusing.
+    ///
+    /// In certain `unsafe` code, lifetime elision combined with
+    /// inconsistent lifetime syntax may result in unsound code.
+    pub MISMATCHED_LIFETIME_SYNTAXES,
+    Warn,
+    "detects when an elided lifetime uses different syntax between arguments and return values"
+}
+
+declare_lint_pass!(LifetimeStyle => [MISMATCHED_LIFETIME_SYNTAXES]);
```

Comment:
```suggestion
declare_lint_pass!(LifetimeSyntaxes => [MISMATCHED_LIFETIME_SYNTAXES]);
```

Probably we want to rename this to match, along with the file/module name.

Reply:
I do not agree with renaming this. The reason is that this module will be gaining additional lints in the followup PR:

```rust
declare_lint_pass!(LifetimeStyle => [
    MISMATCHED_LIFETIME_SYNTAXES,
    HIDDEN_LIFETIMES_IN_INPUT_PATHS,
    HIDDEN_LIFETIMES_IN_OUTPUT_PATHS,
]);
```

Reply:
I can't figure out if there ever was a decision on a lint group name. Or are the three(?) lints just standalone lints?

Reply:
I will wait for a compiler team member to weigh in on the organization of the compiler implementation.

Reply:
AFAICT, this specific lint pass name is compiler internal. I really don't feel like it's materially different (to me, at least) if we call it `LifetimeStyle` or `LifetimeSyntax`, nor is this really worth going back-and-forth on. These lints *are* about the style of the syntaxes (hah) written by the user that causes their program to be possible confusing or misleading to read.

If we're calling `mismatched_lifetime_syntaxes` the lint with a name including "syntaxes", let's just use `LifetimeSyntax` as the lint pass name for now to unblock us (it can be changed later and does not need to be perfect). `HIDDEN_LIFETIMES_IN_INPUT_PATHS` and `HIDDEN_LIFETIMES_IN_OUTPUT_PATHS` from what I can gather is also about style of the syntaxes used.

If we still have no agreement, we can also adopt some lint pass name like `LifetimeSyntaxStyle` that makes nobody happy to unblock us (again, this really is not something I want to block over).